staging/src/k8s.io/apiserver/pkg/reconcilers/lease.go

```diff
@@ -0,0 +1,421 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kruntime "k8s.io/apimachinery/pkg/runtime"
+	endpointsv1 "k8s.io/apiserver/pkg/apis/endpoints"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+	genericfeatures "k8s.io/apiserver/pkg/features"
+	"k8s.io/apiserver/pkg/registry/rest"
+	"k8s.io/apiserver/pkg/storage"
+	"k8s.io/apiserver/pkg/storage/storagebackend"
+	storagefactory "k8s.io/apiserver/pkg/storage/storagebackend/factory"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+)
+
+const (
+	APIServerIdentityLabel = "apiserverIdentity"
+)
+
+// Leases is an interface which assists in managing the set of active masters
+type Leases interface {
+	// ListLeases retrieves a list of the current master IPs
+	ListLeases() ([]string, error)
+
+	// GetLease retrieves the IP of a specific service
+	GetLease(serverId string) (string, error)
+
+	// UpdateLease adds or refreshes a master's lease
+	UpdateLease(ip string, serverId string, endpointPorts []corev1.EndpointPort) error
+
+	// RemoveLease removes a master's lease
+	RemoveLease(ip string, endpointPorts []corev1.EndpointPort) error
+
+	// Destroy cleans up everything on shutdown.
+	Destroy()
+}
+
+type storageLeases struct {
+	storage   storage.Interface
+	destroyFn func()
+	baseKey   string
+	leaseTime time.Duration
+}
+
+var _ Leases = &storageLeases{}
+
+// ListLeases retrieves a list of the current master IPs from storage
+func (s *storageLeases) ListLeases() ([]string, error) {
+	storageOpts := storage.ListOptions{
+		ResourceVersion:      "0",
+		ResourceVersionMatch: metav1.ResourceVersionMatchNotOlderThan,
+		Predicate:            storage.Everything,
+		Recursive:            true,
+	}
+	ipInfoList, err := s.GetIpInfoList(storageOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	ipList := make([]string, 0, len(ipInfoList.Items))
+	for _, ip := range ipInfoList.Items {
+		if len(ip.Subsets) > 0 && len(ip.Subsets[0].Addresses) > 0 && len(ip.Subsets[0].Addresses[0].IP) > 0 {
+			ipList = append(ipList, ip.Subsets[0].Addresses[0].IP)
+		}
+	}
+
+	klog.V(6).Infof("Current master IPs listed in storage are %v", ipList)
+
+	return ipList, nil
+}
+
+// UpdateLease resets the TTL on a master IP in storage
+// UpdateLease will create a new key if it doesn't exist.
+func (s *storageLeases) UpdateLease(ip string, serverId string, endpointPorts []corev1.EndpointPort) error {
+	key := path.Join(s.baseKey, ip)
+	// TODO: how to use endpointPorts to save in the key
+	if len(endpointPorts) > 0 {
+		key = path.Join(key, fmt.Sprint(endpointPorts[0].Port))
+	}
+	return s.storage.GuaranteedUpdate(apirequest.NewDefaultContext(), key, &corev1.Endpoints{}, true, nil, func(input kruntime.Object, respMeta storage.ResponseMeta) (kruntime.Object, *uint64, error) {
+		// just make sure we've got the right IP set, and then refresh the TTL
+		existing := input.(*corev1.Endpoints)
+		existing.Subsets = []corev1.EndpointSubset{
+			{
+				Addresses: []corev1.EndpointAddress{{IP: ip}},
```

Comment:
`Hostname: ID`

```
// EndpointAddress is a tuple that describes single IP address.
// +structType=atomic
type EndpointAddress struct {
	// The IP of this endpoint.
	// May not be loopback (127.0.0.0/8 or ::1), link-local (169.254.0.0/16 or fe80::/10),
	// or link-local multicast (224.0.0.0/24 or ff02::/16).
	IP string `json:"ip" protobuf:"bytes,1,opt,name=ip"`
	// The Hostname of this endpoint
	// +optional
	Hostname string `json:"hostname,omitempty" protobuf:"bytes,3,opt,name=hostname"`
	// Optional: Node hosting this endpoint. This can be used to determine endpoints local to a node.
	// +optional
	NodeName *string `json:"nodeName,omitempty" protobuf:"bytes,4,opt,name=nodeName"`
	// Reference to object providing the endpoint.
	// +optional
	TargetRef *ObjectReference `json:"targetRef,omitempty" protobuf:"bytes,2,opt,name=targetRef"`
}
```